kolibri/auth/test/test_permissions.py

```diff
@@ -1,454 +1,568 @@
+        self.assertFalse(orphan.user_can_read(self.own_classroom_admin))
+        self.assertFalse(orphan.user_can_read(self.own_classroom_coach))
+        self.assertFalse(orphan.user_can_read(self.member))
+
+    def test_facility_user_can_update_self(self):
+        """ A FacilityUser can update its own FacilityUser model """
+        self.assertTrue(self.member.user_can_update(self.member))
+        self.assertTrue(self.own_classroom_coach.user_can_update(self.own_classroom_coach))
+        self.assertTrue(self.own_classroom_admin.user_can_update(self.own_classroom_admin))
+        self.assertTrue(self.data["facilityadmin"].user_can_update(self.data["facilityadmin"]))
+
+    def test_admins_but_not_coaches_can_update_facility_users(self):
+        """ Users with admin (but not coach) role for a FacilityUser can update that FacilityUser """
+        self.assertTrue(self.member.user_can_update(self.data["facilityadmin"]))
+        self.assertFalse(self.member.user_can_update(self.data["facilitycoach"]))
+        self.assertTrue(self.member.user_can_update(self.own_classroom_admin))
+        self.assertFalse(self.member.user_can_update(self.own_classroom_coach))
+
+    def test_admins_and_coaches_for_other_classrooms_cannot_update_facility_users(self):
+        """ Users without admin/coach role for a specific FacilityUser cannot update that FacilityUser """
+        self.assertFalse(self.other_member.user_can_update(self.own_classroom_coach))
+        self.assertFalse(self.other_member.user_can_update(self.own_classroom_admin))
+
+    def test_only_facility_admins_can_update_unaffiliated_facility_users(self):
+        """ Only Facility admins can update FacilityUser that is not a member of a Classroom or LearnerGroup """
+        orphan = self.data["unattached_users"][0]
+        self.assertTrue(orphan.user_can_update(self.data["facilityadmin"]))
+        self.assertFalse(orphan.user_can_update(self.data["facilitycoach"]))
+        self.assertFalse(orphan.user_can_update(self.own_classroom_admin))
+        self.assertFalse(orphan.user_can_update(self.own_classroom_coach))
+        self.assertFalse(orphan.user_can_update(self.member))
+
+    def test_facility_user_cannot_delete_self(self):
+        """ A FacilityUser cannot delete its own FacilityUser model, even for an admin/coach """
+        self.assertFalse(self.member.user_can_delete(self.member))
+        self.assertFalse(self.own_classroom_coach.user_can_delete(self.own_classroom_coach))
+        self.assertFalse(self.own_classroom_admin.user_can_delete(self.own_classroom_admin))
+        self.assertFalse(self.data["facilityadmin"].user_can_delete(self.data["facilityadmin"]))
+
+    def test_only_facility_admins_can_delete_facility_user(self):
+        """ The only FacilityUsers who can delete a FacilityUser are admins for the Facility """
+        self.assertTrue(self.member.user_can_delete(self.data["facilityadmin"]))
```

Comment:
Ah, I see that we're asking whether the user in the parameter has the given permission for the calling user object. However this is a little unexpected to me because Django's permissions are phrased differently. For instance, this LOC might be written as `facilityadmin.can_delete(member)`. This reads more like natural language -- "facility admin can delete member?" in contrast to `member.user_can_delete(facility_admin)` which is ambiguous IMO.

Alternatively we might write `facilityadmin.has_perm("can_delete", member)`, which additionally has the benefit of using the same method for checking all permissions (simply with different parameters) instead of requiring a separate method for every different type of permission.

Do you think it would be worthwhile to use either of these alternative phrasings? Or should we stick to what you've already written here?